File: java/src/main/java/com/myriad/wdl/model/processors/WdlExpressionProcessorBase.java
package com.myriad.wdl.model.processors;

import com.myriad.wdl.model.expressions.WdlArrayLiteral;
import com.myriad.wdl.model.expressions.WdlBinaryOperation;
import com.myriad.wdl.model.expressions.WdlBooleanLiteral;
import com.myriad.wdl.model.expressions.WdlExpression;
import com.myriad.wdl.model.expressions.WdlFloatLiteral;
import com.myriad.wdl.model.expressions.WdlFunctionCallOperation;
import com.myriad.wdl.model.expressions.WdlIndexAccessOperation;
import com.myriad.wdl.model.expressions.WdlIntLiteral;
import com.myriad.wdl.model.expressions.WdlMapLiteral;
import com.myriad.wdl.model.expressions.WdlMapLiteral.WdlMapEntry;
import com.myriad.wdl.model.expressions.WdlMemberAccessOperation;
import com.myriad.wdl.model.expressions.WdlNullLiteral;
import com.myriad.wdl.model.expressions.WdlObjectLiteral;
import com.myriad.wdl.model.expressions.WdlObjectLiteral.WdlObjectEntry;
import com.myriad.wdl.model.expressions.WdlPairLiteral;
import com.myriad.wdl.model.expressions.WdlStringLiteral;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringComponent;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringEscape;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringPlaceholder;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringPlaceholderOption;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringText;
import com.myriad.wdl.model.expressions.WdlStringLiteral.WdlStringToken;
import com.myriad.wdl.model.expressions.WdlStructLiteral;
import com.myriad.wdl.model.expressions.WdlStructLiteral.WdlStructEntry;
import com.myriad.wdl.model.expressions.WdlTernaryOperation;
import com.myriad.wdl.model.expressions.WdlUnaryOperation;
import com.myriad.wdl.model.expressions.WdlVariable;

/**
 * Base depth-first expression walker.
 *
 * <p>Dispatch is done via {@code componentType()} switches rather than {@code instanceof} checks.
 */
public abstract class WdlExpressionProcessorBase implements WdlExpressionProcessor {

  @Override
  public final void processExpression(WdlExpression expression) {
    walkExpression(expression);
  }

  protected final void walkExpression(WdlExpression expression) {
    if (expression == null) {
      return;
    }

    enterExpression(expression);

    switch (expression.componentType()) {
      case BOOL_LIT:
        processBooleanLiteral((WdlBooleanLiteral) expression);
        break;
      case FLOAT_LIT:
        processFloatLiteral((WdlFloatLiteral) expression);
        break;
      case INT_LIT:
        processIntLiteral((WdlIntLiteral) expression);
        break;
      case ARRAY_LIT:
        processArrayLiteral((WdlArrayLiteral) expression);
        break;
      case MAP_LIT:
        processMapLiteral((WdlMapLiteral) expression);
        break;
      case NULL_LIT:
        processNullLiteral((WdlNullLiteral) expression);
        break;
      case OBJ_LIT:
        processObjectLiteral((WdlObjectLiteral) expression);
        break;
      case PAIR_LIT:
        processPairLiteral((WdlPairLiteral) expression);
        break;
      case STR_LIT:
        processStringLiteral((WdlStringLiteral) expression);
        break;
      case STRUCT_LIT:
        processStructLiteral((WdlStructLiteral) expression);
        break;
      case VARIABLE:
        processVariable((WdlVariable) expression);
        break;
      case BINARY_OP:
        processBinaryOperation((WdlBinaryOperation) expression);
        break;
      case FUNC_OP:
        processFunctionCallOperation((WdlFunctionCallOperation) expression);
        break;
      case IDX_OP:
        processIndexAccessOperation((WdlIndexAccessOperation) expression);
        break;
      case MEMBER_OP:
        processMemberAccessOperation((WdlMemberAccessOperation) expression);
        break;
      case TERNARY_OP:
        processTernaryOperation((WdlTernaryOperation) expression);
        break;
      case UNARY_OP:
        processUnaryOperation((WdlUnaryOperation) expression);
        break;
      default:
        throw new IllegalStateException(
            "Unhandled expression component type: " + expression.componentType());
    }

    exitExpression(expression);
  }

  protected final void walkStringComponent(WdlStringLiteral context, WdlStringComponent component) {
    if (component == null) {
      return;
    }

    enterStringComponent(context, component);

    switch (component.componentType()) {
      case TEXT:
        processStringText(context, (WdlStringText) component);
        break;
      case ESC:
        processStringEscape(context, (WdlStringEscape) component);
        break;
      case SPECIAL:
        processStringToken(context, (WdlStringToken) component);
        break;
      case PLACEHOLDER:
        processStringPlaceholder(context, (WdlStringPlaceholder) component);
        break;
      default:
        throw new IllegalStateException(
            "Unhandled string component type: " + component.componentType());
    }

    exitStringComponent(context, component);
  }

  protected void enterExpression(WdlExpression expression) {}

  protected void exitExpression(WdlExpression expression) {}

  protected void enterStringComponent(WdlStringLiteral context, WdlStringComponent component) {}

  protected void exitStringComponent(WdlStringLiteral context, WdlStringComponent component) {}

  protected void processBooleanLiteral(WdlBooleanLiteral expression) {}

  protected void processFloatLiteral(WdlFloatLiteral expression) {}

  protected void processIntLiteral(WdlIntLiteral expression) {}

  protected void processNullLiteral(WdlNullLiteral expression) {}

  protected void processVariable(WdlVariable expression) {}

  protected void processArrayLiteral(WdlArrayLiteral expression) {
    expression.entries().forEach(this::walkExpression);
  }

  protected void processMapLiteral(WdlMapLiteral expression) {
    for (WdlMapEntry entry : expression.entries()) {
      processMapEntry(expression, entry);
    }
  }

  protected void processMapEntry(WdlMapLiteral context, WdlMapEntry entry) {
    walkExpression(entry.getKey());
    walkExpression(entry.getValue());
  }

  protected void processObjectLiteral(WdlObjectLiteral expression) {
    for (WdlObjectEntry entry : expression.entries()) {
      processObjectEntry(expression, entry);
    }
  }

  protected void processObjectEntry(WdlObjectLiteral context, WdlObjectEntry entry) {
    walkExpression(entry.getValue());
  }

  protected void processPairLiteral(WdlPairLiteral expression) {
    walkExpression(expression.getLeft());
    walkExpression(expression.getRight());
  }

  protected void processStringLiteral(WdlStringLiteral expression) {
    for (WdlStringComponent component : expression.components()) {
      walkStringComponent(expression, component);
    }
  }

  protected void processStringText(WdlStringLiteral context, WdlStringText text) {}

  protected void processStringEscape(WdlStringLiteral context, WdlStringEscape escape) {}

  protected void processStringToken(WdlStringLiteral context, WdlStringToken token) {}

  protected void processStringPlaceholder(
      WdlStringLiteral context, WdlStringPlaceholder placeholder) {
    processStringPlaceholderOption(context, placeholder, placeholder.getOption());
    walkExpression(placeholder.getExpression());
  }

  protected void processStringPlaceholderOption(
      WdlStringLiteral context,
      WdlStringPlaceholder placeholder,
      WdlStringPlaceholderOption option) {
    if (option == null) {
      return;
    }

    switch (option.getType()) {
      case SEP:
      case DEFAULT:
        walkExpression(option.getValue());
        break;
      case TRUE_FALSE:
      case FALSE_TRUE:
        walkExpression(option.getTrueValue());
        walkExpression(option.getFalseValue());
        break;
      default:
        throw new IllegalStateException("Unhandled placeholder option type: " + option.getType());
    }
  }

  protected void processStructLiteral(WdlStructLiteral expression) {
    for (WdlStructEntry entry : expression.entries()) {
      processStructEntry(expression, entry);
    }
  }

  protected void processStructEntry(WdlStructLiteral context, WdlStructEntry entry) {
    walkExpression(entry.getValue());
  }

  protected void processBinaryOperation(WdlBinaryOperation expression) {
    walkExpression(expression.getLeft());
    walkExpression(expression.getRight());
  }

  protected void processFunctionCallOperation(WdlFunctionCallOperation expression) {
    expression.arguments().forEach(this::walkExpression);
  }

  protected void processIndexAccessOperation(WdlIndexAccessOperation expression) {
    walkExpression(expression.getTarget());
    walkExpression(expression.getIndex());
  }

  protected void processMemberAccessOperation(WdlMemberAccessOperation expression) {
    walkExpression(expression.getTarget());
  }

  protected void processTernaryOperation(WdlTernaryOperation expression) {
    walkExpression(expression.getCondition());
    walkExpression(expression.getTrueValue());
    walkExpression(expression.getFalseValue());
  }

  protected void processUnaryOperation(WdlUnaryOperation expression) {
    walkExpression(expression.getOperand());
  }
}

File: python/src/wdl_model/model/processors/wdl_expression_processor_base.py
"""Depth-first expression walker for the Python WDL object model."""

from __future__ import annotations

from wdl_model.model.expressions import (
    WdlStringPlaceholderOptionType,
    WdlArrayLiteral,
    WdlBinaryOperation,
    WdlBooleanLiteral,
    WdlExpression,
    WdlFloatLiteral,
    WdlFunctionCallOperation,
    WdlIndexAccessOperation,
    WdlIntLiteral,
    WdlMapEntry,
    WdlMapLiteral,
    WdlMemberAccessOperation,
    WdlNullLiteral,
    WdlObjectEntry,
    WdlObjectLiteral,
    WdlPairLiteral,
    WdlStringComponent,
    WdlStringEscape,
    WdlStringLiteral,
    WdlStringPlaceholder,
    WdlStringPlaceholderOption,
    WdlStringText,
    WdlStringToken,
    WdlStructEntry,
    WdlStructLiteral,
    WdlTernaryOperation,
    WdlUnaryOperation,
    WdlVariable,
)

from .wdl_expression_processor import WdlExpressionProcessor


class WdlExpressionProcessorBase(WdlExpressionProcessor):
    """Base depth-first expression walker."""

    def processExpression(self, expression: WdlExpression | None) -> None:
        self.walkExpression(expression)

    def walkExpression(self, expression: WdlExpression | None) -> None:
        if expression is None:
            return

        self.enterExpression(expression)

        if isinstance(expression, WdlBooleanLiteral):
            self.processBooleanLiteral(expression)
        elif isinstance(expression, WdlFloatLiteral):
            self.processFloatLiteral(expression)
        elif isinstance(expression, WdlIntLiteral):
            self.processIntLiteral(expression)
        elif isinstance(expression, WdlArrayLiteral):
            self.processArrayLiteral(expression)
        elif isinstance(expression, WdlMapLiteral):
            self.processMapLiteral(expression)
        elif isinstance(expression, WdlNullLiteral):
            self.processNullLiteral(expression)
        elif isinstance(expression, WdlObjectLiteral):
            self.processObjectLiteral(expression)
        elif isinstance(expression, WdlPairLiteral):
            self.processPairLiteral(expression)
        elif isinstance(expression, WdlStringLiteral):
            self.processStringLiteral(expression)
        elif isinstance(expression, WdlStructLiteral):
            self.processStructLiteral(expression)
        elif isinstance(expression, WdlVariable):
            self.processVariable(expression)
        elif isinstance(expression, WdlBinaryOperation):
            self.processBinaryOperation(expression)
        elif isinstance(expression, WdlFunctionCallOperation):
            self.processFunctionCallOperation(expression)
        elif isinstance(expression, WdlIndexAccessOperation):
            self.processIndexAccessOperation(expression)
        elif isinstance(expression, WdlMemberAccessOperation):
            self.processMemberAccessOperation(expression)
        elif isinstance(expression, WdlTernaryOperation):
            self.processTernaryOperation(expression)
        elif isinstance(expression, WdlUnaryOperation):
            self.processUnaryOperation(expression)
        else:
            raise RuntimeError(
                f"Unhandled expression type: {expression.__class__.__name__}"
            )

        self.exitExpression(expression)

    def walkStringComponent(
        self,
        context: WdlStringLiteral,
        component: WdlStringComponent | None,
    ) -> None:
        if component is None:
            return

        self.enterStringComponent(context, component)
        if isinstance(component, WdlStringText):
            self.processStringText(context, component)
        elif isinstance(component, WdlStringEscape):
            self.processStringEscape(context, component)
        elif isinstance(component, WdlStringToken):
            self.processStringToken(context, component)
        elif isinstance(component, WdlStringPlaceholder):
            self.processStringPlaceholder(context, component)
        else:
            raise RuntimeError(
                f"Unhandled string component type: {component.__class__.__name__}"
            )
        self.exitStringComponent(context, component)

    def enterExpression(self, expression: WdlExpression) -> None:
        pass

    def exitExpression(self, expression: WdlExpression) -> None:
        pass

    def enterStringComponent(
        self, context: WdlStringLiteral, component: WdlStringComponent
    ) -> None:
        pass

    def exitStringComponent(
        self, context: WdlStringLiteral, component: WdlStringComponent
    ) -> None:
        pass

    def processBooleanLiteral(self, expression: WdlBooleanLiteral) -> None:
        pass

    def processFloatLiteral(self, expression: WdlFloatLiteral) -> None:
        pass

    def processIntLiteral(self, expression: WdlIntLiteral) -> None:
        pass

    def processNullLiteral(self, expression: WdlNullLiteral) -> None:
        pass

    def processVariable(self, expression: WdlVariable) -> None:
        pass

    def processArrayLiteral(self, expression: WdlArrayLiteral) -> None:
        for e in expression.entries():
            self.walkExpression(e)

    def processMapLiteral(self, expression: WdlMapLiteral) -> None:
        for entry in expression.entries():
            self.processMapEntry(expression, entry)

    def processMapEntry(self, context: WdlMapLiteral, entry: WdlMapEntry) -> None:
        self.walkExpression(entry.getKey())
        self.walkExpression(entry.getValue())

    def processObjectLiteral(self, expression: WdlObjectLiteral) -> None:
        for entry in expression.entries():
            self.processObjectEntry(expression, entry)

    def processObjectEntry(
        self, context: WdlObjectLiteral, entry: WdlObjectEntry
    ) -> None:
        self.walkExpression(entry.getValue())

    def processPairLiteral(self, expression: WdlPairLiteral) -> None:
        self.walkExpression(expression.left)
        self.walkExpression(expression.right)

    def processStringLiteral(self, expression: WdlStringLiteral) -> None:
        for component in expression.components():
            self.walkStringComponent(expression, component)

    def processStringText(self, context: WdlStringLiteral, text: WdlStringText) -> None:
        pass

    def processStringEscape(
        self, context: WdlStringLiteral, escape: WdlStringEscape
    ) -> None:
        pass

    def processStringToken(
        self, context: WdlStringLiteral, token: WdlStringToken
    ) -> None:
        pass

    def processStringPlaceholder(
        self, context: WdlStringLiteral, placeholder: WdlStringPlaceholder
    ) -> None:
        self.processStringPlaceholderOption(context, placeholder, placeholder.option)
        self.walkExpression(placeholder.expression)

    def processStringPlaceholderOption(
        self,
        context: WdlStringLiteral,
        placeholder: WdlStringPlaceholder,
        option: WdlStringPlaceholderOption | None,
    ) -> None:
        if option is None:
            return
        if option.type == WdlStringPlaceholderOptionType.DEFAULT:
            self.walkExpression(option.value)
        elif option.type == WdlStringPlaceholderOptionType.TRUE_FALSE:
            self.walkExpression(option.trueValue)
            self.walkExpression(option.falseValue)
        else:
            raise RuntimeError(f"Unhandled placeholder option type: {option.type}")

    def processStructLiteral(self, expression: WdlStructLiteral) -> None:
        for entry in expression.entries():
            self.processStructEntry(expression, entry)

    def processStructEntry(
        self, context: WdlStructLiteral, entry: WdlStructEntry
    ) -> None:
        self.walkExpression(entry.getValue())

    def processBinaryOperation(self, expression: WdlBinaryOperation) -> None:
        self.walkExpression(expression.left)
        self.walkExpression(expression.right)

    def processFunctionCallOperation(
        self, expression: WdlFunctionCallOperation
    ) -> None:
        for arg in expression.arguments():
            self.walkExpression(arg)

    def processIndexAccessOperation(self, expression: WdlIndexAccessOperation) -> None:
        self.walkExpression(expression.target)
        self.walkExpression(expression.index)

    def processMemberAccessOperation(
        self, expression: WdlMemberAccessOperation
    ) -> None:
        self.walkExpression(expression.target)

    def processTernaryOperation(self, expression: WdlTernaryOperation) -> None:
        self.walkExpression(expression.condition)
        self.walkExpression(expression.trueValue)
        self.walkExpression(expression.falseValue)

    def processUnaryOperation(self, expression: WdlUnaryOperation) -> None:
        self.walkExpression(expression.operand)
